packages/Meshfree/src/DTK_MovingLeastSquaresOperator_def.hpp

```diff
@@ -0,0 +1,133 @@
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+
+#ifndef DTK_MOVING_LEAST_SQUARES_OPERATOR_DEF_HPP
+#define DTK_MOVING_LEAST_SQUARES_OPERATOR_DEF_HPP
+
+#include <DTK_DBC.hpp>
+#include <DTK_DetailsMovingLeastSquaresOperatorImpl.hpp>
+#include <DTK_DetailsNearestNeighborOperatorImpl.hpp> // makeDistributedSearchTree, fetch
+#include <DTK_DistributedSearchTree.hpp>
+
+namespace DataTransferKit
+{
+
+template <typename DeviceType, typename CompactlySupportedRadialBasisFunction,
+          typename PolynomialBasis>
+MovingLeastSquaresOperator<DeviceType, CompactlySupportedRadialBasisFunction,
+                           PolynomialBasis>::
+    MovingLeastSquaresOperator(
+        Teuchos::RCP<Teuchos::Comm<int> const> const &comm,
+        Kokkos::View<Coordinate const **, DeviceType> source_points,
+        Kokkos::View<Coordinate const **, DeviceType> target_points,
+        Teuchos::ParameterList const &plist )
+    : _comm( comm )
+    , _offset( "offset" )
+    , _ranks( "ranks" )
+    , _indices( "indices" )
+    , _coeffs( "polynomial_coefficients" )
+{
+    double const radius = plist.get<double>( "radius" );
+
+    DTK_REQUIRE( source_points.extent_int( 1 ) ==
+                 target_points.extent_int( 1 ) );
+    // FIXME for now let's assume 3D
+    DTK_REQUIRE( source_points.extent_int( 1 ) == 3 );
+
+    // Build distributed search tree over the source points.
+    auto search_tree = Details::NearestNeighborOperatorImpl<
+        DeviceType>::makeDistributedSearchTree( _comm, source_points );
+    DTK_CHECK( !search_tree.empty() );
+
+    // Query points within a given radius for all target points.
+    auto queries = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::makeWithinRadiusQueries( target_points, radius );
+
+    // Perform the actual search.
+    search_tree.query( queries, _indices, _offset, _ranks );
+
+    // Retrieve the coordinates of all source points that met the predicates.
+    // NOTE: This is the last collective.
+    source_points = Details::NearestNeighborOperatorImpl<DeviceType>::fetch(
+        _comm, _ranks, _indices, source_points );
+
+    // TODO: ensure that we found enough source points to proceed
+
+    // Transform source points
+    source_points = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::transformSourceCoordinates( source_points, _offset,
+                                                 target_points );
+    target_points = Kokkos::View<double **, DeviceType>( "empty" );
+
+    // Build P (vandermonde matrix)
+    // P is a single 1D storage for multiple P_i matrices. Each matrix is of
+    // size (#source_points_for_specific_target_point, basis_size)
+    auto p =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeVandermonde(
+            source_points, PolynomialBasis() );
+
+    // Build phi (weight matrix)
```

Comment:
Let's leave the implementation this way for now but it will be interesting to see if we get performance differences by building `p`, `phi`, 'a', and other intermediates in order to build `_coeffs` rather than having sequences of `KOKKOS_INLINE_FUNCTION` calls that execute the entire coefficient sequence in a single `parallel_for`. I think this is similar to what `Intrepid2` decided to do as well.